torch/_functorch/aot_autograd.py

```diff
@@ -1648,6 +1648,52 @@ def same_dtype_views(view1, view2):
     return True
 
 
+
+def tensors_definitely_do_not_overlap(x, y):
+    if torch._C._are_definitely_non_overlapping(x, y):
+        # not overlapping (this covers the basic cases that eager mode uses)
+        return True
+
+    # expected inputs: two lists of intervals in increasing order. Don't need to have the same length
+    def intervals_intersect(xs, ys):
+        if len(xs) == 0 or len(ys) == 0:
+            return False
+        x_ = xs[0]
+        y_ = ys[0]
+        if max(0, min(x_[1], y_[1]) - max(x_[0], y_[0])) > 0:
+            return True
+        if x_[1] > y_[1]:
+            return intervals_intersect(xs, ys[1:])
+        else:
+            return intervals_intersect(xs[1:], ys)
+
+    # This cases is needed for the shampoo optimizer.
+    # We can compute a list of contiguous intervals and check if any of them intersect
+    if x.dim() == 2 and y.dim() == 2 and x.stride()[1] == 1 and y.stride()[1] == 1:
+        # Compute the (start, end) range of contiguous memory of each region of the tensor
+        x_ranges = [
```

Comment:
my code is under the assumption that `x.stride(0) == y.stride(0)` for sure! But that seems to be the case you care about anyways?

Reply:
Talked offline - agreed to add the faster check you mentioned to the C++ code (which does not have that check today, since it bails out if either tensor is noncontiguous)